### PR TITLE
Enhance filesystem.process_file get_metadata cache performance.   

### PR DESCRIPTION
**Pull request by amaleki**
Did some more profiling on the new master  https://github.com/jmathai/elodie/commit/d8cee15f32d3853070111c972ecf987a05a55549 with over 500 images and noticed that calls to pyexiftool (get_metadata) are double the file count.  Meaning each image processed was making two calls to get_metadata.  It turns out in filesystem (process_file) there is a media.set_original_name sandwiched between two media.get_metadata.  All media setters have a reset_cache method that clears the cache.  I've moved the media.set_original_name after the last media.get_metadata to ensure all getters are using cache.     

image

image

After making this change you can see base.py (get_metadata) drops from 55% to 37% of runtime.   Also the number of calls to pyexiftool (get_metadata) drop by almost half.  

image

image

One side effect of this change is the time executing "utime" increases from 200ms to 7 seconds.  It's not clear to me yet why this is the case or if its just an anomaly.
**Master**
image
**Pull Request**
image

It's a simple change but could have unintended consequence.  Appreciate if you could take a look at it.
Here are the prof files:
[elodie-amaleki-process-file-setter.zip](https://github.com/jmathai/elodie/files/4063609/elodie-amaleki-process-file-setter.zip)

### CODE REVIEW

**Comment by amaleki** [comment]
Moving the media.set_original_name caused problems.  

Other option that seems to work is changing last **media.get_metadata** to reference **metadata** variable.  It passed unit tests and the files look good.  Same time savings and call reduction as previous commit.  

image

image

File: elodie/filesystem.py [hunk 1]
```diff
@@ -589,7 +589,7 @@ def process_file(self, _file, destination, media, **kwargs):
             #  before we made any changes.
             # Then set the utime on the destination file based on metadata.
             os.utime(_file, (stat_info_original.st_atime, stat_info_original.st_mtime))
-            self.set_utime_from_metadata(media.get_metadata(), dest_path)
+            self.set_utime_from_metadata(metadata, dest_path)
```
Comment (jmathai) [inline elodie/filesystem.py]: Looks like this is a safe change. Calling `.get_metadata()` was part of a [PR](https://github.com/jmathai/elodie/pull/331/files) where `.set_original_name()` was added and prior to that commit `metadata` wasn't stored as a variable.

**Comment by jmathai** [comment]
Thanks for this PR. It looks safe but I'm not seeing much performance improvement when running tests. Current master and this PR branch execute in around 1 min 20 sec.

[master](https://travis-ci.org/jmathai/elodie/builds/636748989)
[pull-request](https://travis-ci.org/jmathai/elodie/builds/637757618)

**Comment by amaleki** [comment]
You will only see improvement in a long running elodie import.  Example above is for 500+ images.  I'll open up an issue to ask the community to run some benchmarks.  I will also do some further benchmarks.  I run elodie on a synology DS718+ in a docker container.

**Comment by Jogai** [comment]
I vote for merging this PR. Suppose the call is going to be more expensive in the future, then the difference would be bigger. I just finished importing 140000 files before all the perf changes from @amaleki . I would have liked every improvement. 
And why not, it an improvement regardless...

**Comment by amaleki** [comment]
Here is a real world test.  I ran 2 tests on each master and PR353 with image/movie files on a Synology DS718+ running Elodie in a docker container running Alpine Linux.  Volumes are mounted to the docker container to access image directories.    

### Summary
~4465 files 21.0GB  - **master:** 19m32s **PR353:** 16m32s **(15.35% decrease)**
~9566 files 56.1GB - **master:** 53m5s **PR353:** 44m52s **(15.47% decrease)**

### System Specifications
Synology DS718+ 
Intel Celeron J3455
16GB RAM
2 X 4TB WD RED Drives

Elodie is run in a docker container
Linux elodie-pr353 4.4.59+ #24922 SMP PREEMPT Mon Aug 19 12:13:37 CST 2019 x86_64 Linux

### ELODIE-MASTER
time ./elodie.py import --destination="./processed/" --source="./incoming/5000/"
Metric      Count                                        
Success      4465                                                                                       
Error           9
                                                                                                                            
**real    19m32.013s**                                        
user    9m59.116s                                       
sys     1m52.070s   

time ./elodie.py import --destination="./processed/" --source="./incoming/10000/"
Metric      Count                                                            
Success      9566                                         
Error           9 
                                                                                                                                             
**real    53m4.953s**                                  
user    26m39.693s                                        
sys     5m19.024s 


### ELODIE PR353
time ./elodie.py import --destination="./processed/" --source="./incoming/5000/"                        
**Metric      Count**                                                                              
Success      4465                                     
Error           9    
                                                                                                                                            
**real    16m32.968s**                                         
user    6m12.309s                                        
sys     1m48.689s   

time ./elodie.py import --destination="./processed/" --source="./incoming/10000/"
**Metric      Count**                                                          
Success      9566                      
Error           9                         
                           
**real    44m52.013s**                         
user    17m29.652s                       
sys     5m19.401s

**Comment by jmathai** [comment]
Thanks @amaleki for the continued performance enhancements :)

**Pull request merged**